### PR TITLE
Reuse decompressed woff files based on hash

### PR DESCRIPTION
**Pull request by ColinKinloch**
When adding a woff font to a `FontConfiguration` only decode it if it is new.

Brings my build times down from 140s to 20s.

Sorry about #1495 and #1500. I didn't realise that the fonts filename was being used by font config as an unique id for its properties.

### CODE REVIEW

**Review by liZe — CHANGES_REQUESTED** [review]
Thanks a lot for the PR!

Could you please fix the couple of problems listed in the comments? Apart from that, the PR looks good to me, we can merge it for version 54.

File: weasyprint/text/fonts.py [hunk 1]
```diff
@@ -194,17 +196,27 @@ def add_font_face(self, rule_descriptors, url_fetcher):
...
-                        font = out.getvalue()
+                    font_is_woff = font[:3] == b'wOF'
+                    if font_is_woff:
+                        font_hasher = hashlib.sha256()
+                        font_hasher.update(font)
+                        font_digest = font_hasher.hexdigest()
+                        decoded_woff_path = self._woff_cache.get(font_digest)
+                        if decoded_woff_path is None:
+                            out = io.BytesIO()
+                            if font[3:4] == b'F':
```
[the remark this answers is not in the document]
Comment (ColinKinloch) [reply weasyprint/text/fonts.py:207]: It seems font[3] returns a number, font[3:4] returns a bytes

File: weasyprint/text/fonts.py [hunk 2]
```diff
@@ -194,17 +196,27 @@ def add_font_face(self, rule_descriptors, url_fetcher):
...
+                                # woff font
+                                ttfont = TTFont(io.BytesIO(font))
+                                ttfont.flavor = ttfont.flavorData = None
+                                ttfont.save(out)
+                            elif font[3:4] == b'2':
+                                # woff2 font
+                                woff2.decompress(io.BytesIO(font), out)
+                            font = out.getvalue()
+                        else:
+                            fd = open('rb', decoded_woff_path)
```
[the remark this answers is not in the document]
Comment (ColinKinloch) [reply weasyprint/text/fonts.py]: :facepalm: I've added a second woff font-face to the test.

**Review by liZe — APPROVED** [review]

**Comment by liZe** [comment]
Thank you!

**Pull request merged**